### DISCUSSION TITLE
How to start  background_tasks  on event  startup

### DISCUSSION THREAD

**Discussion by HuRuWo** [fastapi/fastapi#8854] [Questions] [answered]
```
@app.on_event("startup")
async def startup_event(background_tasks: BackgroundTasks):
    send_notification(background_tasks)
```

```
2021-05-12 15:31:40,350 - MainThread - ERROR - Traceback (most recent call last):
  File "F:\douyin_spider_fast_api\venv\lib\site-packages\starlette\routing.py", line 540, in lifespan
    async for item in self.lifespan_context(app):
  File "F:\douyin_spider_fast_api\venv\lib\site-packages\starlette\routing.py", line 481, in default_lifespan
    await self.startup()
  File "F:\douyin_spider_fast_api\venv\lib\site-packages\starlette\routing.py", line 516, in startup
    await handler()
TypeError: startup_event() missing 1 required positional argument: 'background_tasks'
```

**Comment by Hultner** [comment]
Where are your background_tasks defined? The problem right now is that you are expecting it as an argument for the startup_event which is called upon by the lifecycle handler without this argument. You'd have to retrieve this information from somewhere.

You might be interested in the [EasySchedule](https://github.com/codemation/easyschedule) library by @codemation which handles task scheduling in a more fine grained manner.

**Comment by tusharjagtap-bc** [comment]
Any update on this? I'm hitting the same issue.

**Comment by codemation** [comment]
@tusharjagtap-bc this is a non-issue. 

on_event('startup') does not expect the decorated method to have arguments. Ask yourself, why are arguments required in this task?

**Comment by tusharjagtap-bc** [comment]
@codemation That is an issue in my opinion. on_event('startup') does not mean that you will always want to call a method without arguments on startup, does it? 

In any case, a workaround to be able to start a background task on startup is highly appreciated. I haven't found a way to do so yet.

**Comment by codemation** [comment]
The most obvious way to achieve referencing your method that takes arguments, is with a new method with the arguments statically defined: 

```
@app.on_event('startup')
async def startup_event_setup():
    startup_event(background_tasks)
```

@tusharjagtap-bc It might help if you share what argument that you are passing into your startup event.

**Comment by Hultner** [comment]
As codemation said, you must have these background_tasks referenced somewhere, so calling the desired function from within the event handler makes the most sense. You could also provide it as a default argument but I don't see much reason to do that.

**Comment by stur86** [comment]
> As codemation said, you must have these background_tasks referenced somewhere, so calling the desired function from within the event handler makes the most sense. You could also provide it as a default argument but I don't see much reason to do that.

No, this is incorrect. In regular entry points like `.get`, `background_tasks` simply gets injected automatically by FastAPI. You don't have to instantiate it outside. So you can't in fact solve this problem that way. Perhaps the hack that might do it is to set up a path that launches the background task and then have an `on_startup` event that sends a request to the app itself to kick it off. But it's silly that one has to resort to such means.

**Comment by Leohanhart** [comment]
Giant fix: Create a deamon thread, create a script that needs to runs in the background. Create a startup_event, trigger the state function like the policy assigns, run the thread, never look back. 

USING THE BUILD IN BACKGROUNDTASKS ONSTARTUP IS A !@#%& DISASTER. 


Just saying, 22-05-2022 now. Im adding a background task on startup, just the way I would expect it. 
I followed instruction, I made a static function, added a startup_event. Tried it on any way. 
    
@app.on_event("startup")
async def startup_event(background_tasks: BackgroundTasks):
    
    #tried it as startup_script.start_update_scedule()
    startup_event(background_tasks.add_task(onstart_function))
    # tried it as background_tasks.add_task(onstart_function)
 
and nothing worked. This is a serious issue here, make it !$@%! work. 
Prettending this problem is not an real problem - even if it is brought to you mulitple times - makes me rember why I never hire  !@#!% narcists. Start making clear policy's on how to start a !@#%^^& background task when the program starts.

**Comment by jayvynl** [comment, marked as answer]
Here is a duplicated question and this is the answer https://github.com/tiangolo/fastapi/issues/2713#issuecomment-768949823

**Comment by humblepie** [comment]
I would like a dependency injection solution for BackgroundTasks in startup_event to be consistent with how the route handlers work.  However, maybe BackgroundTasks was only meant for running a background task as part of an API request/response cycle.

I decided to just implement my long running task in a separate thread when the app startup event triggers instead of trying to hook into fastapi's BackgroundTasks
